README.md
<div align="center">
  
# 🖼️💾 Electron Window State

[![Node CI](https://github.com/BetaHuhn/electron-win-state/workflows/Node%20CI/badge.svg)](https://github.com/BetaHuhn/electron-win-state/actions?query=workflow%3A%22Node+CI%22) [![Release CI](https://github.com/BetaHuhn/electron-win-state/workflows/Release%20CI/badge.svg)](https://github.com/BetaHuhn/electron-win-state/actions?query=workflow%3A%22Release+CI%22) [![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/BetaHuhn/electron-win-state/blob/master/LICENSE) ![David](https://img.shields.io/david/betahuhn/electron-win-state)

Store and restore your Electron Window's Size and Position.

</div>

## Features

- 💾 **Persistent** - *persistently stores your Electron Window's size and position*
- 🔌 **Easy integration** - *integrates easily with your existing BrowserWindow configuration*
- 🔨 **Customization** - *customize the behaviour and the underlying [electron-store](https://github.com/sindresorhus/electron-store) instance*
- 🖼️ **Frame option** - *you can [optionally](storing-the-frame-option) store the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option as well*

## 🚀 Get started

```shell
npm install electron-win-state
```

*Requires Electron 11 or later*

## 📚 Usage

The easiest way to use [electron-win-state](https://github.com/BetaHuhn/electron-win-state), is to add it as a wrapper around your normal [`BrowserWindow`](https://www.electronjs.org/docs/api/browser-window):

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	width: 800,
	height: 600,
	// your normal BrowserWindow options...
})
```

This will create a new [`BrowserWindow`](https://www.electronjs.org/docs/api/browser-window) with the provided options and store any changes once the window is closed.

Thats it! Your Electron app now has a persistent window! 🎉

### Manually

You can also use [electron-win-state](https://github.com/BetaHuhn/electron-win-state) manually:

```js
const winState = new WinState({ 
	defaultWidth: 800,
	defaultHeight: 600,
	// other winState options, see below
})

const browserWindow = new BrowserWindow({
	...winState.winOptions,
	// your normal BrowserWindow options...
})

// Attach the required event listeners
winState.manage(this.browserWindow)
```

With `winState.winOptions` you have access to the restored size and position of the window: 

```js
console.log(winState.winOptions) // => { width: 800, height: 600, x: 0, y: 0, frame: true }
```

## ⚙️ Options

You can customize the behaviour of [electron-win-state](https://github.com/BetaHuhn/electron-win-state) further, by passing an options object to `new WinState()` or by specifying a `winState` object when using `.createBrowserWindow()`:

```js
new WinState({ 
	defaultWidth: 800,
	defaultHeight: 600,
	dev: true,
	addMethods: true
})
```

Or:

```js
WinState.createBrowserWindow({
    winState: {
		defaultWidth: 800,
		defaultHeight: 600,
		dev: true,
		addMethods: true
	}
})
```

Or:

```js
WinState.createBrowserWindow({
	width: 800,
	height: 600,
	winState: {
		dev: true,
		addMethods: true
	}
})
```

All of these result in the same.

Here are all the options [electron-win-state](https://github.com/BetaHuhn/electron-win-state) supports:
	
| Name | Type | Description | Default |
| ------------- | ------------- | ------------- | ------------- |
| `defaultWidth` | `number` | The default width which will be used when no stored value was found | `800` |
| `defaultHeight` | `number` | The default height which will be used when no stored value was found | `600` |
| `defaultFrame` | `boolean` | The default value for the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option when no stored value was found | `true` |
| `storeFrameOption` | `boolean` | Store and restore the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option (will be enabled automatically if `defaultFrame` is provided) | `false` |
| `dev` | `boolean` | Enable development mode. Changes will be stored immediately after resizing or moving and not just after closing a window | `false` |
| `addMethods` | `boolean` | Add the `.resetWindowToDefault()`, `.setFramed()` and `.getStoredWinOptions()` methods to the provided [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) | `true` |
| `electronStoreOptions` | [`object`](https://github.com/sindresorhus/electron-store#options) | Will be passed to the underlying [electron-store](https://github.com/sindresorhus/electron-store) instance | `{ name: 'window-state' }` |
| `store` | [`instance`](https://github.com/sindresorhus/electron-store#instance) | An existing [electron-store](https://github.com/sindresorhus/electron-store) instance to use | n/a |

## 📖 Examples

Here are a few examples to help you get started!

### Basic Example

This is the most basic example. A new [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) will be created and the required event listeners will be attached automatically.

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	width: 800,
	height: 600
})

browserWindow.loadURL('https://github.com/BetaHuhn/electron-win-state')
```

---

### Development Mode

During development it might be helpful to store the window size and position immediately after changing it and not just after closing the window. You can enable this functionality by setting `dev` to true:

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	width: 800,
	height: 600,
	winState: {
		dev: true
	}
})

browserWindow.loadURL('https://github.com/BetaHuhn/electron-win-state')
```

### Resetting the Window

If `addMethods` is enabled (it is by default) a `.resetWindowToDefault()` method will be added to the provided [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) and can be used to both reset the stored state, as well as resizing the window to it's defaults:

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	width: 800,
	height: 600,
	winState: {
		dev: true
	}
})

// Later
browserWindow.resetWindowToDefault()
```

This method could also be accessed anywhere else in your Electron app by getting the currently focused window:

```js
const browserWindow = BrowserWindow.getFocusedWindow()
browserWindow.resetWindowToDefault()
```

---

### Storing the frame option

You can optionally store the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option as well. This might be useful if you want your users to enable or disable the window frame and store their choice.

If `storeFrameOption` is enabled (or `defaultFrame` is provided) the `frame` or `defaultFrame` option will be stored and can be later changed with the `.setFramed()` method on the provided [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) (will be added if `addMethods` is true).

> You have to recreate the window for the `frame` option to take effect

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	width: 800,
	height: 600,
	frame: false,
	winState: {
		storeFrameOption: true
	}
})

// Later
browserWindow.setFramed(true)
```

This method could also be accessed anywhere else in your Electron app by getting the currently focused window:

```js
const browserWindow = BrowserWindow.getFocusedWindow()
browserWindow.setFramed(true)
```

---

### Get stored values

You can access the restored window size and position with the `winOptions` object:

```js
const winState = new WinState({ 
	defaultWidth: 800,
	defaultHeight: 600,
	defaultFrame: false
})

winState.winOptions // => { width: 1200, height: 700, x: 50, y: 105, frame: true }
```

You can get the stored values at any time using the added `.getStoredWinOptions()` method on the [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) (only added if `addMethods` is true):

```js
import WinState from 'electron-win-state'

const browserWindow = WinState.createBrowserWindow({
	winState: {
		defaultWidth: 800,
		defaultHeight: 600,
		defaultFrame: false
	}
})

browserWindow.getStoredWinOptions() // => { width: 1200, height: 700, x: 50, y: 105, frame: true }
```

---

## 💻 Development

Issues and PRs are very welcome!

- run `yarn lint` or `npm run lint` to run eslint.
- run `yarn watch` or `npm run watch` to watch for changes.
- run `yarn build` or `npm run build` to produce a compiled version in the `lib` folder.

## ❔ About

This project was developed by me ([@betahuhn](https://github.com/BetaHuhn)) in my free time. If you want to support me:

[![Donate via PayPal](https://img.shields.io/badge/paypal-donate-009cde.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=394RTSBEEEFEE)

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/F1F81S2RK)

### Credit

This library uses the great [electron-store](https://github.com/sindresorhus/electron-store) by [@sindresorhus](https://github.com/sindresorhus) under the hood and was inspired by [electron-window-state](https://github.com/mawie81/electron-window-state) and [electron-boilerplate](https://github.com/szwacz/electron-boilerplate/blob/master/src/helpers/window.js).

## 📄 License

Copyright 2021 Maximilian Schiller

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
## [v1.1.22] - 2021-11-15

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.22) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.21...v1.1.22) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.22) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.22.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.22.tar.gz))

### Dependency updates

- [`b0a277a`](https://github.com/BetaHuhn/electron-win-state/commit/b0a277a)  Bump electron from 15.3.0 to 15.3.1

## [v1.1.21] - 2021-10-25

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.21) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.20...v1.1.21) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.21) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.21.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.21.tar.gz))

### Dependency updates

- [`b41ccd9`](https://github.com/BetaHuhn/electron-win-state/commit/b41ccd9)  Bump electron from 12.0.10 to 12.1.0

## [v1.1.20] - 2021-10-23

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.20) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.19...v1.1.20) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.20) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.20.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.20.tar.gz))

### Dependency updates

- [`962ec63`](https://github.com/BetaHuhn/electron-win-state/commit/962ec63)  Bump electron from 12.0.18 to 12.1.0

## [v1.1.19] - 2021-10-18

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.19) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.18...v1.1.19) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.19) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.19.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.19.tar.gz))

### Dependency updates

- [`90fcbec`](https://github.com/BetaHuhn/electron-win-state/commit/90fcbec)  Bump typescript from 4.4.3 to 4.4.4

## [v1.1.18] - 2021-10-11

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.18) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.17...v1.1.18) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.18) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.18.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.18.tar.gz))

### Dependency updates

- [`4b2c978`](https://github.com/BetaHuhn/electron-win-state/commit/4b2c978)  Bump @typescript-eslint/eslint-plugin from 4.32.0 to 4.33.0
- [`9d3eca9`](https://github.com/BetaHuhn/electron-win-state/commit/9d3eca9)  Bump @typescript-eslint/parser from 4.32.0 to 4.33.0

## [v1.1.17] - 2021-10-04

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.17) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.16...v1.1.17) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.17) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.17.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.17.tar.gz))

### Dependency updates

- [`b47a240`](https://github.com/BetaHuhn/electron-win-state/commit/b47a240)  Bump @typescript-eslint/eslint-plugin from 4.31.2 to 4.32.0
- [`f987944`](https://github.com/BetaHuhn/electron-win-state/commit/f987944)  Bump @typescript-eslint/parser from 4.31.2 to 4.32.0

## [v1.1.16] - 2021-09-27

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.16) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.15...v1.1.16) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.16) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.16.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.16.tar.gz))

### Dependency updates

- [`ecb678a`](https://github.com/BetaHuhn/electron-win-state/commit/ecb678a)  Bump @types/debounce from 1.2.0 to 1.2.1
- [`4b9abf6`](https://github.com/BetaHuhn/electron-win-state/commit/4b9abf6)  Bump @typescript-eslint/eslint-plugin from 4.31.1 to 4.31.2
- [`b47d42d`](https://github.com/BetaHuhn/electron-win-state/commit/b47d42d)  Bump @typescript-eslint/parser from 4.31.1 to 4.31.2

## [v1.1.15] - 2021-09-20

[Release notes](https://github.com/BetaHuhn/electron-win-state/releases/tag/v1.1.15) · [Compare](https://github.com/BetaHuhn/electron-win-state/compare/v1.1.14...v1.1.15) · [Tag](https://github.com/BetaHuhn/electron-win-state/tree/v1.1.15) · Archive ([zip](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.15.zip) · [tar.gz](https://github.com/BetaHuhn/electron-win-state/archive/v1.1.15.tar.gz))

### Dependency updates

- [`0689c40`](https://github.com/BetaHuhn/electron-win-state/commit/0689c40)  Bump typescript from 4.4.2 to 4.4.3
- [`f0c4a21`](https://github.com/BetaHuhn/electron-win-state/commit/f0c4a21)  Bump @typescript-eslint/eslint-plugin from 4.31.0 to 4.31.1
- [`25c4ab1`](https://github.com/BetaHuhn/electron-win-state/commit/25c4ab1)  Bump @typescript-eslint/parser from 4.31.0 to 4.31.1
- [`a3db3ce`](https://github.com/BetaHuhn/electron-win-state/commit/a3db3ce)  Bump electron-store from 8.0.0 to 8.0.1

## [v1.1.14] - 2021-09-13


lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debounce_1 = __importDefault(require("debounce"));
const electron_store_1 = __importDefault(require("electron-store"));
const electron_1 = require("electron");
/**
 * Store and restore your Electron Window's Size and Position.
 * @example
    ```
    const winState = new WinState({
        defaultWidth: 800,
        defaultHeight: 600,
        // other winState options, see below
    })

    const browserWindow = new BrowserWindow({
        ...winState.winOptions,
        // your normal BrowserWindow options...
    })

    // Attach the required event listeners
    winState.manage(this.browserWindow)
    ```
 */
class WinState {
    constructor(options) {
        const defaultOptions = {
            defaultWidth: 600,
            defaultHeight: 800,
            defaultFrame: true,
            storeFrameOption: options.defaultFrame !== undefined || false,
            dev: false,
            debounce: 500,
            electronStoreOptions: {
                name: 'window-state'
            },
            addMethods: true,
            store: undefined
        };
        this.opts = Object.assign({}, defaultOptions, options);
        this.store = this.opts.store ? this.opts.store : new electron_store_1.default(this.opts.electronStoreOptions);
        this.state = this.getState();
    }
    getState() {
        const stored = this.store.store;
        const defaults = Object.assign({ width: this.opts.defaultWidth, height: this.opts.defaultHeight }, (this.opts.storeFrameOption && { frame: this.opts.defaultFrame }));
        return Object.assign({}, defaults, stored);
    }
    saveState() {
        this.store.set(this.state);
    }
    /**
     * Change the stored frame option
     *
     * Note: You need to recreate the window for this to take effect
     */
    changeFrame(value) {
        this.state.frame = value;
        this.saveState();
    }
    /**
     * Reset the stored window size and position
     */
    reset() {
        this.store.clear();
    }
    /**
     * Attach event listeners to the BrowserWindow.
     *
     * Will listen to the `resize`, `move`, `close` and `closed event`.
     *
     * By default the changes will only be stored on the `close` and `closed` event.
     * Use the `dev` option to store the changes on `resize` and `move` as well.
     * @param win
     */
    manage(win) {
        this.win = win;
        this.win.on('resize', (0, debounce_1.default)(() => this.changeHandler(), this.opts.debounce));
        this.win.on('move', (0, debounce_1.default)(() => this.changeHandler(), this.opts.debounce));
        this.win.on('close', () => this.closeHandler());
        this.win.on('closed', () => this.closeHandler());
        // Add a reset method to the window
        if (this.opts.addMethods) {
            this.win.resetWindowToDefault = () => {
                var _a;
                (_a = this.win) === null || _a === void 0 ? void 0 : _a.setSize(this.opts.defaultWidth, this.opts.defaultHeight);
                this.reset();
            };
            this.win.setFramed = (value) => {
                this.changeFrame(value);
            };
            this.win.getStoredWinOptions = () => {
                return this.winOptions;
            };
        }
    }
    /**
     * Remove all attached event listeners
     */
    unmanage() {
        if (this.win) {
            this.win.removeListener('resize', (0, debounce_1.default)(() => this.changeHandler(), this.opts.debounce));
            this.win.removeListener('move', (0, debounce_1.default)(() => this.changeHandler(), this.opts.debounce));
            this.win.removeListener('close', () => this.closeHandler());
            this.win.removeListener('closed', () => this.closeHandler());
            this.win = undefined;
        }
    }
    changeHandler() {
        try {
            if (!this.win)
                return;
            const winBounds = this.win.getBounds();
            if (this.isNormal()) {
                this.state.x = winBounds.x;
                this.state.y = winBounds.y;
                this.state.width = winBounds.width;
                this.state.height = winBounds.height;
            }
            // Not working, reference: https://git.io/JZ3n5
            /* this.state.isMaximized = this.win.isMaximized()
            this.state.isFullScreen = this.win.isFullScreen() */
            if (this.opts.dev) {
                this.saveState();
            }
        }
        catch (err) {
            // Don't throw an error when window was closed
        }
    }
    closeHandler() {
        // Unregister listeners and save state
        this.unmanage();
        this.saveState();
    }
    isNormal() {
        var _a, _b, _c;
        return !((_a = this.win) === null || _a === void 0 ? void 0 : _a.isMaximized()) && !((_b = this.win) === null || _b === void 0 ? void 0 : _b.isMinimized()) && !((_c = this.win) === null || _c === void 0 ? void 0 : _c.isFullScreen());
    }
    /**
     * Create a new [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) with the restored window size and position.
     *
     * Will attach the event listeners automatically.
     * @param options Options for the new BrowserWindow as well as [electron-win-state](https://github.com/BetaHuhn/electron-win-state) itself.
     * @returns A new BrowserWindow
     * @example
        ```
        import WinState from 'electron-win-state'

        const browserWindow = WinState.createBrowserWindow({
            width: 800,
            height: 600,
            // your normal BrowserWindow options...
        })
        ```
     */
    static createBrowserWindow(options) {
        var _a, _b, _c;
        // If storeFrameOption is missing and the defaultFrame property exists enable storeFrameOption
        if (((_a = options.winState) === null || _a === void 0 ? void 0 : _a.storeFrameOption) === undefined) {
            options.winState = Object.assign(Object.assign({}, options.winState), { storeFrameOption: ((_b = options.winState) === null || _b === void 0 ? void 0 : _b.defaultFrame) !== undefined });
        }
        // Parse winState specific options from options
        const winStateOpts = Object.assign({}, Object.assign({ defaultWidth: options.width, defaultHeight: options.height }, (((_c = options.winState) === null || _c === void 0 ? void 0 : _c.storeFrameOption) && { defaultFrame: options.frame })), options.winState);
        const winState = new WinState(winStateOpts);
        // Cleanup options object
        delete options.winState;
        // Create a new BrowserWindow with the provided options and the current winState
        const win = new electron_1.BrowserWindow(Object.assign(Object.assign({}, options), winState.winOptions));
        winState.manage(win);
        return win;
    }
    /**
     * The current window size and position
     */
    get winOptions() {
        return Object.assign(Object.assign({ width: this.width, height: this.height }, (this.opts.storeFrameOption && { frame: this.frame })), { x: this.x, y: this.y });
    }
    /**
     * The current window width
     */
    get width() {
        return this.state.width;
    }
    /**
     * The current window height
     */
    get height() {
        return this.state.height;
    }
    /**
     * If the window has a frame
     */
    get frame() {
        return this.state.frame;
    }
    /**
     * The current window x position
     */
    get x() {
        return this.state.x;
    }
    /**
     * The current window y position
     */
    get y() {
        return this.state.y;
    }
}
exports.default = WinState;


src/index.ts
import debounce from 'debounce'
import Store from 'electron-store'
import { BrowserWindow } from 'electron'

import { Options, FinalOptions, State, CreateBrowserWindowOptions } from './types'

/**
 * Store and restore your Electron Window's Size and Position.
 * @example
	```
	const winState = new WinState({ 
		defaultWidth: 800,
		defaultHeight: 600,
		// other winState options, see below
	})

	const browserWindow = new BrowserWindow({
		...winState.winOptions,
		// your normal BrowserWindow options...
	})

	// Attach the required event listeners
	winState.manage(this.browserWindow)
	```
 */
export default class WinState<T> {

	opts: FinalOptions<T>
	store: Store
	state: State
	win?: BrowserWindow

	constructor(options: Options<T>) {
		const defaultOptions = {
			defaultWidth: 600,
			defaultHeight: 800,
			defaultFrame: true,
			storeFrameOption: options.defaultFrame !== undefined || false,
			dev: false,
			debounce: 500,
			electronStoreOptions: { 
				name: 'window-state'
			},
			addMethods: true,
			store: undefined
		}

		this.opts = Object.assign({}, defaultOptions, options)
		this.store = this.opts.store ? this.opts.store : new Store(this.opts.electronStoreOptions as any)
		this.state = this.getState()
	}

	getState() {
		const stored = this.store.store
		const defaults = { width: this.opts.defaultWidth, height: this.opts.defaultHeight, ...(this.opts.storeFrameOption && { frame: this.opts.defaultFrame }) }

		return Object.assign({}, defaults, stored)
	}

	saveState() {
		this.store.set(this.state as any)
	}

	/**
	 * Change the stored frame option
	 * 
	 * Note: You need to recreate the window for this to take effect
	 */
	changeFrame(value: boolean) {
		this.state.frame = value

		this.saveState()
	}

	/**
	 * Reset the stored window size and position
	 */
	reset() {
		this.store.clear()
	}

	/**
	 * Attach event listeners to the BrowserWindow.
	 * 
	 * Will listen to the `resize`, `move`, `close` and `closed event`.
	 * 
	 * By default the changes will only be stored on the `close` and `closed` event.
	 * Use the `dev` option to store the changes on `resize` and `move` as well.
	 * @param win 
	 */
	manage(win: BrowserWindow) {
		this.win = win

		this.win.on('resize', debounce(() => this.changeHandler(), this.opts.debounce))
		this.win.on('move', debounce(() => this.changeHandler(), this.opts.debounce))
		this.win.on('close', () => this.closeHandler())
		this.win.on('closed', () => this.closeHandler())

		// Add a reset method to the window
		if (this.opts.addMethods) {
			(this.win as any).resetWindowToDefault = () => {
				this.win?.setSize(this.opts.defaultWidth, this.opts.defaultHeight)
				this.reset()
			}

			(this.win as any).setFramed = (value: boolean) => {
				this.changeFrame(value)
			}

			(this.win as any).getStoredWinOptions = () => {
				return this.winOptions
			}
		}
	}

	/**
	 * Remove all attached event listeners
	 */
	unmanage() {
		if (this.win) {
			this.win.removeListener('resize', debounce(() => this.changeHandler(), this.opts.debounce))
			this.win.removeListener('move', debounce(() => this.changeHandler(), this.opts.debounce))
			this.win.removeListener('close', () => this.closeHandler())
			this.win.removeListener('closed', () => this.closeHandler())
			this.win = undefined
		}
	}

	changeHandler() {
		try {
			if (!this.win) return

			const winBounds = this.win.getBounds()

			if (this.isNormal()) {
				this.state.x = winBounds.x
				this.state.y = winBounds.y
				this.state.width = winBounds.width
				this.state.height = winBounds.height
			}

			// Not working, reference: https://git.io/JZ3n5
			/* this.state.isMaximized = this.win.isMaximized()
			this.state.isFullScreen = this.win.isFullScreen() */

			if (this.opts.dev) {
				this.saveState()
			}
		} catch (err) {
			// Don't throw an error when window was closed
		}
	}

	closeHandler() {
		// Unregister listeners and save state
		this.unmanage()
		this.saveState()
	}

	isNormal() {
		return !this.win?.isMaximized() && !this.win?.isMinimized() && !this.win?.isFullScreen()
	}

	/**
	 * Create a new [BrowserWindow](https://www.electronjs.org/docs/api/browser-window) with the restored window size and position.
	 * 
	 * Will attach the event listeners automatically.
	 * @param options Options for the new BrowserWindow as well as [electron-win-state](https://github.com/BetaHuhn/electron-win-state) itself.
	 * @returns A new BrowserWindow
	 * @example
		```
		import WinState from 'electron-win-state'

		const browserWindow = WinState.createBrowserWindow({
			width: 800,
			height: 600,
			// your normal BrowserWindow options...
		})
		```
	 */
	static createBrowserWindow(options: CreateBrowserWindowOptions<any>): BrowserWindow {
		// If storeFrameOption is missing and the defaultFrame property exists enable storeFrameOption
		if (options.winState?.storeFrameOption === undefined) {
			options.winState = {
				...options.winState,
				storeFrameOption: options.winState?.defaultFrame !== undefined
			}
		}

		// Parse winState specific options from options
		const winStateOpts = Object.assign({}, { defaultWidth: options.width, defaultHeight: options.height, ...(options.winState?.storeFrameOption && { defaultFrame: options.frame }) }, options.winState)

		const winState = new WinState(winStateOpts)

		// Cleanup options object
		delete options.winState

		// Create a new BrowserWindow with the provided options and the current winState
		const win = new BrowserWindow({ ...options, ...winState.winOptions })

		winState.manage(win)

		return win
	}

	/**
	 * The current window size and position
	 */
	get winOptions() {
		return {
			width: this.width,
			height: this.height,
			...(this.opts.storeFrameOption && { frame: this.frame }),
			x: this.x,
			y: this.y
		}
	}

	/**
	 * The current window width
	 */
	get width() {
		return this.state.width
	}

	/**
	 * The current window height
	 */
	get height() {
		return this.state.height
	}

	/**
	 * If the window has a frame
	 */
	 get frame() {
		return this.state.frame
	}

	/**
	 * The current window x position
	 */
	get x() {
		return this.state.x
	}

	/**
	 * The current window y position
	 */
	get y() {
		return this.state.y
	}
}

src/types.ts
import Store, { Options as electronStoreOptions } from 'electron-store'
import { SetRequired } from 'type-fest'
import { BrowserWindowConstructorOptions } from 'electron'

/**
 * Options to configure [electron-win-state](https://github.com/BetaHuhn/electron-win-state)
 */
export interface Options<T> {
	/**
	 * Default width which will be used when no stored value was found
	 * 
	 * @default 800
	 */
	defaultWidth?: number;

	/**
	 * Default height which will be used when no stored value was found
	 * 
	 * @default 600
	 */
	defaultHeight?: number;

	/**
	 * Default value for the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option
	 * 
	 * @default true
	 */
	defaultFrame?: boolean;

	/**
	 * Store and restore the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option
	 * 
	 * Will be enabled automatically if `defaultFrame` is provided.
	 * @default false
	 */
	storeFrameOption?: boolean;

	/**
	 * Enable development mode
	 * 
	 * Changes will be stored immediately after resizing or moving and not just after closing a window.
	 * 
	 * @default false
	 */
	dev?: boolean;

	/**
	 * Add the `.resetWindowToDefault()`, `.setFramed()` and `.getStoredWinOptions()` methods to the provided [BrowserWindow](https://www.electronjs.org/docs/api/browser-window)
	 * 
	 * Can be used to both reset the stored state, as well as resizing the window to it's defaults.
	 * 
	 * @default true
	 */
	addMethods?: boolean;

	debounce?: number;
	electronStoreOptions?: electronStoreOptions<T>;
	store?: Store;
}

export type FinalOptions<T> = SetRequired<Options<T>, 'defaultWidth' | 'defaultHeight' | 'defaultFrame' | 'storeFrameOption' | 'dev' | 'debounce' | 'addMethods' | 'electronStoreOptions'>

export interface CreateBrowserWindowOptions<T> extends BrowserWindowConstructorOptions {
	/**
	 * Default width which will be used when no stored value was found
	 * 
	 * @default 800
	 */
	width?: number;

	/**
	 * Default height which will be used when no stored value was found
	 * 
	 * @default 600
	 */
	height?: number;

	/**
	 * Options to configure [electron-win-state](https://github.com/BetaHuhn/electron-win-state) itself
	 */
	winState?: Options<T>;
}

export interface State {
	/**
	 * The current window width
	 */
	width: number;

	/**
	 * The current window height
	 */
	height: number;

	/**
	 * If the window has a frame
	 */
	frame?: boolean;

	/**
	 * The current window x position
	 */
	x?: number;

	/**
	 * The current window y position
	 */
	y?: number;

	/**
	 * Indicates if the window is maximized
	 */
	isMaximized?: boolean;

	/**
	 * Indicates if the window is in fullscreen
	 */
	isFullScreen?: boolean;
}

lib/types.d.ts
import Store, { Options as electronStoreOptions } from 'electron-store';
import { SetRequired } from 'type-fest';
import { BrowserWindowConstructorOptions } from 'electron';
/**
 * Options to configure [electron-win-state](https://github.com/BetaHuhn/electron-win-state)
 */
export interface Options<T> {
    /**
     * Default width which will be used when no stored value was found
     *
     * @default 800
     */
    defaultWidth?: number;
    /**
     * Default height which will be used when no stored value was found
     *
     * @default 600
     */
    defaultHeight?: number;
    /**
     * Default value for the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option
     *
     * @default true
     */
    defaultFrame?: boolean;
    /**
     * Store and restore the [frame](https://www.electronjs.org/docs/api/frameless-window#frameless-window) option
     *
     * Will be enabled automatically if `defaultFrame` is provided.
     * @default false
     */
    storeFrameOption?: boolean;
    /**
     * Enable development mode
     *
     * Changes will be stored immediately after resizing or moving and not just after closing a window.
     *
     * @default false
     */
    dev?: boolean;
    /**
     * Add the `.resetWindowToDefault()`, `.setFramed()` and `.getStoredWinOptions()` methods to the provided [BrowserWindow](https://www.electronjs.org/docs/api/browser-window)
     *
     * Can be used to both reset the stored state, as well as resizing the window to it's defaults.
     *
     * @default true
     */
    addMethods?: boolean;
    debounce?: number;
    electronStoreOptions?: electronStoreOptions<T>;
    store?: Store;
}
export declare type FinalOptions<T> = SetRequired<Options<T>, 'defaultWidth' | 'defaultHeight' | 'defaultFrame' | 'storeFrameOption' | 'dev' | 'debounce' | 'addMethods' | 'electronStoreOptions'>;
export interface CreateBrowserWindowOptions<T> extends BrowserWindowConstructorOptions {
    /**
     * Default width which will be used when no stored value was found
     *
     * @default 800
     */
    width?: number;
    /**
     * Default height which will be used when no stored value was found
     *
     * @default 600
     */
    height?: number;
    /**
     * Options to configure [electron-win-state](https://github.com/BetaHuhn/electron-win-state) itself
     */
    winState?: Options<T>;
}
export interface State {
    /**
     * The current window width
     */
    width: number;
    /**
     * The current window height
     */
    height: number;
    /**
     * If the window has a frame
     */
    frame?: boolean;
    /**
     * The current window x position
     */
    x?: number;
    /**
     * The current window y position
     */
    y?: number;
    /**
     * Indicates if the window is maximized
     */
    isMaximized?: boolean;
    /**
     * Indicates if the window is in fullscreen
     */
    isFullScreen?: boolean;
}
